[RNMobile] Try fixing split/merge issue on native mobile by dropping selection update event when late

hypest opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

## Description
This is another attempt to fix https://github.com/WordPress/gutenberg/issues/29478. Opening this set of PRs to have a view of what CI thinks.

There's [this prior PR already](https://github.com/WordPress/gutenberg/pull/29654) but since @jd-alexander noticed that the issue still happens sometimes, I gave it a different attempt. We'll collaborate on which one to keep if any.

gutenberg-mobile PR: https://github.com/wordpress-mobile/gutenberg-mobile/pull/3239

## How has this been tested?
Manual tests of splitting and merging paragraph blocks. Also, we ran the ["Writing flow" section of the manual testsuite](https://github.com/wordpress-mobile/test-cases/tree/trunk/test-cases/gutenberg/writing-flow) for native mobile.

## Types of changes
* Started bumping the event counter number on the JS side to denote a need to force update the native side. The RichText component bumps the `lastEventCount` (only for Android, the iOS case still uses `undefined` to pass the same "message") and the native Aztec wrapper component uses that number to synchronize the counter it keeps internally. Note, the JS side doesn't directly change the native counter; the sync happens in native to ensure that there's no race condition happening by two entities trying to change the same counter variable.
* Started dropping the events that reach the JS side from Aztec but are late.

### Individual changes in the native Aztec wrapper
1. [capture the new event counter sent from the JS side](https://github.com/WordPress/gutenberg/pull/29683/files#diff-18c4abf20356408efc28d1d949b45f1b8f479898f849eb0af84e538ebd9322f4R221)
1. In native Aztec wrapper, [sync from the JS side counter when the native component tries to increment the counter](https://github.com/WordPress/gutenberg/pull/29683/files#diff-05b836896093785229e8db7c5b56c130cb98102706d9948068d6d4c378d0fe8eR390). Essentially, instead of just incrementing the native counter, syncing it first and then incrementing.

### Individual changes in the RichText component on the JS side
1. Replace the `this.lastEventCount = undefined` with `this.manipulateEventCounterToForceNativeToRefresh()` as the way to force the native **Android** component to update. It was needed to change this for all occurrences otherwise there were issues in various writing flow situations.
1. [drop the `onChangeFromAztec` event if it's late](https://github.com/WordPress/gutenberg/pull/29683/files#diff-7af18df9c3e07e67518894550a2e1d9d627feb879a2003a2e231ef927d4d06f8R294).
1. [drop the `handleDelete` event if late](https://github.com/WordPress/gutenberg/pull/29683/files#diff-7af18df9c3e07e67518894550a2e1d9d627feb879a2003a2e231ef927d4d06f8R378).
1. Introduce the `shouldDropEventFromAztec` function that [checks if the event counter of an event is late compared to the counter kept locally](https://github.com/WordPress/gutenberg/pull/29683/files#diff-7af18df9c3e07e67518894550a2e1d9d627feb879a2003a2e231ef927d4d06f8R609) in RichText. Only for Android though. For iOS the function always returns false as the counter based dropping mechanism doesn't exist on iOS, doesn't seem to be needed anyway, yet at least.
1. [drop the `onSelectionChangeFromAztec` event if late](https://github.com/WordPress/gutenberg/pull/29683/files#diff-7af18df9c3e07e67518894550a2e1d9d627feb879a2003a2e231ef927d4d06f8R621). This is the main fix for the issue at hand actually, since that event is holding a version of the text before the split, and we don't want to use that text.
1. Introduce the [`manipulateEventCounterToForceNativeToRefresh()` function, that bumps the event counter for Android](https://github.com/WordPress/gutenberg/pull/29683/files#diff-7af18df9c3e07e67518894550a2e1d9d627feb879a2003a2e231ef927d4d06f8R699). Using this function simplifies the callsites trying to trigger the force update of the native side.


## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/native-mobile.md -->

hypest committed:
Bump native event count and drop selection update event if late

hypest committed:
Limit the event count bump and event dropping to iOS

jd-alexander reviewed (commented):
Hi @hypest thanks for these changes. I tested the behavior and it works most of the time but similar to my solution, the bug can still occur even though the solution resolves the issue. Upon observing the content changes of the editor more closely I am realizing that after around 10 splits/merges the bug only occurs if a space is not in front of the word the caret is positioned at before enter is pressed. I am wondering if this has something to do with how the newline marker is being processed in the `EnterPressedWatcher` I am trying to determine a mechanism for proper diagnosis.

hypest commented:
Thanks for giving this a test @jd-alexander !

> after around 10 splits/merges the bug only occurs if a space is not in front of the word the caret is positioned at before enter is pressed

Hmm, can you try capturing a video perhaps of the situation?
I've tried to split+merge in the middle of a word (so to not have spaces around the caret) many times in a row and I can't reproduce on my Pixel 2XL Android 10.

hypest commented:
Noticed that the solution as is introduces a bug:
1. Split a paragraph and merge it again (e.g. by backspacing)
2. Move the caret to another position in the paragraph by tapping on the new position
3. Notice the logcat/console message `Dropping onSelectionChange from Aztec as its event counter is older than latest sent to the native side`. It shouldn't need to happen as the new caret positioning is intentional.
4. Try to split the paragraph in the new caret location. The split will actually happen in the position it happened in step 1 💥

Need to work some more on this.

jd-alexander commented:
> Thanks for giving this a test @jd-alexander !
> 
> > after around 10 splits/merges the bug only occurs if a space is not in front of the word the caret is positioned at before enter is pressed
> 
> Hmm, can you try capturing a video perhaps of the situation? Also, just to cover various cases, does that happen to you both on physical devices and on emulators?
> 
> I've tried to split+merge in the middle of a word (so to not have spaces around the caret) many times in a row and I can't reproduce on my Pixel 2XL Android 10.

Understood! I am going to attempt to reproduce and share a video from my device. IIRC, I observed the behavior on my emulator so my reproduction attempt now will give me a solid idea of the bug's occurrence. 

So the steps I have been using are. 

1. Paste content into a paragraph block. 
2. Put the caret in the middle of the content directly in front of a word. 
3. Press enter. 
4. Put the caret at the end of the first word in the newly created paragraph block that has the second part of the split content. 
5. Press backspace until the block merges. 
6. Repeat steps 2 - 5 consecutively until the content duplication is observed.

hypest committed:
Sync native mobile event counter to the JS side one

This way, when the JS bumps the counter, the native side picks it up and
increments from there.

hypest commented:
Added https://github.com/WordPress/gutenberg/pull/29683/commits/b20609cbeecc227deade9a3f97ca5ff868508993 to sync the two sides (JS, native) to avoid dropping all those events that started be marked as late. Generally this can be a risky change, leading to potential race conditions or bugs in places not really foreseen... this will need good testing.

hypest commented:
> I observed the behavior on my emulator

Actually @jd-alexander , can you share the emulator details too? Device frame, API level, CPU/ABI, with Play Store or not? Thanks!

hypest commented:
OK, I think I replicated the issue on the emulator, here are the steps, looks like I can replicate consistently in the Pixel 3a API 11 emulator **but not on the physical device**:

1. Use WPAndroid
2. use this text in a post: <details><summary>Example paragaph text</summary>The full potential and creativity of any Automattician won’t be completely reflected in their assigned tasks or OKRs, or documented in the issue or task tracker a team might use. It’s possible to survive by doing just what’s asked of you, but to truly thrive involves bringing something above and beyond to your work and making sure everything you leave your mark on is something you’re proud of. We’re a distributed company with 1,398 Automatticians in 79 countries speaking 99 different languages. We’re committed to diversity, equity, and inclusion, and our common goal is to democratize publishing and commerce so that anyone with a story can tell it, and anyone with a product can sell it, regardless of income, gender, politics, language, or where they live in the world.</details>
2. Place the cursor at the middle of the paragraph, in front of the start of a word.
3. Hit "Enter" on the virtual keyb to split the paragraph
4. Place the caret in the middle of the _second_ word in the new paragraph
5. Hit backspace (virtual keyb) until the blocks are merged
6. Continue hitting backspace until a word is deleted
7. Hit Enter to split the block again
8. Hit backspace to merge back (without any other char deletions)
9. 💥 text from the split paragraph is duplicated before the cursor

Video capture of the steps and issue: 

https://user-images.githubusercontent.com/1032913/110615733-65ef6080-819c-11eb-9384-b2169ff99c27.mp4

hypest commented:
Didn't have the chance to work on debugging/fixing not that there seems to be a consistent way to replicate. I did try to use more `manipulateEventCounterToForceNativeToRefresh` in place of `this.lastEventCount = undefined` to remove occurances of `Tried to bump the RichText native event counter but was 'undefined'. Aborting bump.`, but that didn't resolve the issue. @jd-alexander feel free to pick up as well as I'm not going to be able to work on this today, need to be at a half-day training thing, thanks!

hypest committed:
More usage of event counter bumping instead of undefining

hypest committed:
Introducing shouldDropEventFromAztec and using it in onChange too

So to drop the Aztec onChange events too, if late.

hypest committed:
Don't bump the local event counter on componentDidUpdate

If the update was due to content change from the JS side of things, the
counter would have already been bumped in shouldComponentUpdate
(hopefully), so bumping it again in DidUpdate introduces a counter bump
that gets out-of-sync with Aztec since (the bump inside DidUpdate
doesn't cause a render and thus no new counter sent to Aztec).

hypest commented:
Added a couple more commits to fix [the issue](https://github.com/WordPress/gutenberg/pull/29683#issuecomment-795216123).

The bit that kinda worries me and am not super sure it won't introduce some regression is the [removal of the counter manipulation in `componentDidUpdate`](https://github.com/WordPress/gutenberg/pull/29683/commits/629d81f5e0ea88f1847a0ce05ddb5fd0602125b7). History of that need is buried behind the monorepo super-squash and some code/file movements prior that, making it hard to find what's the rationale/need of that manipulation in the first place. Code archeology needs more time to uncover that I guess 😬.

In the meantime, I'll resort to the testuites and manual testing to verify if some regression is introduced.

jd-alexander commented:
> Added a couple more commits to fix [the issue](https://github.com/WordPress/gutenberg/pull/29683#issuecomment-795216123).

Thanks! I will check this out!
 
> 
> The bit that kinda worries me and am not super sure it won't introduce some regression is the [removal of the counter manipulation in `componentDidUpdate`](https://github.com/WordPress/gutenberg/pull/29683/commits/629d81f5e0ea88f1847a0ce05ddb5fd0602125b7). History of that need is buried behind the monorepo super-squash and some code/file movements prior that, making it hard to find what's the rationale/need of that manipulation in the first place. Code archeology needs more time to uncover that I guess 😬.

Exactly! Every time I try to find some history, the mono repo super squash gets in the way, so I have to utilize the debugger or search for previous PRs to create some form of history. 
> 
> In the meantime, I'll resort to the testuites and manual testing to verify if some regression is introduced. Edit: the mobile testsuites here and in the gutenberg-mobile PR are green 🎉.

 I appreciate the easier reproduction steps mentioned above. 🙏🏾

jd-alexander commented:
I created a [test](https://github.com/wordpress-mobile/WordPress-Android/pull/14239) PR that generates an APK to test these changes.

jd-alexander commented:
### Concern 

> The bit that kinda worries me and am not super sure it won't introduce some regression is the removal of the counter manipulation in componentDidUpdate.
 
Understood! I will investigate this! 

### Investigation

> History of that need is buried behind the monorepo super-squash and some code/file movements prior that, making it hard to find what's the rationale/need of that manipulation in the first place. Code archeology needs more time to uncover that I guess 😬.

True. I did a bit of git sleuthing to find the introduction of the `componentDidUpdate` behavior in the `RichText` component. 

1. As you have said, I observed that the `RichText` component was located in `packages/block-editor/src/components/rich-text/index.native.js` but was later moved to `https://github.com/WordPress/gutenberg/blob/trunk/packages/rich-text/src/component/index.native.js` so to find the PR that added the change, I sifted through changes in `packages/block-editor/src/components/rich-text/` package. 

2. Originally, the behavior was defined in `[RN Mobile] Fix crash when merging block` https://github.com/WordPress/gutenberg/pull/15392 At the time of it's implementation `componentWillReceiveProps` was being utilized. 

https://github.com/WordPress/gutenberg/blob/50527d3120000b5fe874af6ac5fb1826edcaef62/packages/block-editor/src/components/rich-text/index.native.js#L612

3. This [commit](https://github.com/WordPress/gutenberg/pull/15392/commits/3bd6a25f0413ff821b1962481c342a478bd78405) introduced the logic. `componentWillReceiveProps` was marked as unsafe in React 17, so a PR that brought some of the latest splitting changes from the web, and did a [migration](https://github.com/WordPress/gutenberg/pull/15762) of the behavior from `componentWillReceiveProps` to `componentDidUpdate` in this [commit](https://github.com/WordPress/gutenberg/pull/15762/commits/341b484b9350d71e61c52ac1157444c3983d61cc) . 

-----------

### Testing

Device: Google Pixel 4 XL
OS: Android 11
App: WPAndroid
Build: [99043](https://github.com/wordpress-mobile/WordPress-Android/pull/14239)

> In the meantime, I'll resort to the testuites and manual testing to verify if some regression is introduced. Edit: the mobile testsuites here and in the gutenberg-mobile PR are green 🎉.

This is a good testing strategy. In addition to this, I used the testing strategy of the `gutenberg-mobile` [PR](https://github.com/wordpress-mobile/gutenberg-mobile/pull/949) that introduced the changes to verify no regressions were introduced.

I also did the splitting and merging writing flow tests. 

#### Splitting and merging

  - [x] TC001 - Merge after writing
  - [x] TC002 - Merge after selection
  - [x] TC003 - Merge after deleting text
  - [x] TC004 - Merge after deleting all
  - [x] TC005 - Merge multiple blocks
  - [x] TC006 - Splitting/merge list block

## Issues

### Splitting/merge list block 🟡 

When testing, I noticed odd behavior while trying to split the list at times. I was able to reproduce this in other versions of the app like `16.9` & `16.8`, so I don't think these changes are causing this behavior. Is this a regression?

**How to reproduce it:**
1. Write some items on a list block.
2. Place the caret at the end of different items and press enter to split.
3. Notice that sometimes it does not work as expected and causes wonky behavior.

<details><summary><strong>Example</strong></summary>


https://user-images.githubusercontent.com/1509205/110866945-f5c40600-8293-11eb-8eab-6e5609e24653.mp4


</details>

### Next steps

1. Verify that the changes introduced do not contribute to the splitting issues. 
2. Create a test iOS PR and test these changes.

hypest commented:
> This commit introduced the logic. componentWillReceiveProps was marked as unsafe in React 17, so a PR that brought some of the latest splitting changes from the web, and did a migration of the behavior from componentWillReceiveProps to componentDidUpdate in this commit .

Thanks for digging Joel, good find. Yeah, it's not clear why the set-to-undefined was exactly needed there. I'd say let's assume that it's no longer needed if all our testing works fine, especially since there's no crash introduced here (that original PR of Tug's was fixing a crash).

> I also did the splitting and merging writing flow tests.

Nice to see the split/merge manual tests pass, thanks Joel!

> When testing, I noticed odd behavior while trying to split the list at times. I was able to reproduce this in other versions of the app like 16.9 & 16.8, so I don't think these changes are causing this behavior. Is this a regression?

Possibly a regression yea, we should look at it separately from this PR. Haven't tried myself but looks like not a blocker for the PR.

hypest commented:
### Android Writing Flow checklist

Device: Pixel 2XL
OS: Android 11
Build: RP1A.201005.004.A1

#### General
- [ ] TC001 - Paste formatted text copied from website
- [ ] TC001 - Multiline components
#### Rich Text Format
  - [ ] TC001 - Bold, Italic, strikethrough buttons
  - [ ] TC002 - Alignment buttons
  - [ ] TC003 - Alignment Split
  - [ ] TC004 - Link button works without selection
  - [ ] TC005 - Link button works with a selected word
  - [ ] TC006 - Adding a link from a copied URL
  - [ ] TC007 - Test format detection under the cursor
  - [ ] TC008 - Test formatting doesn't remove leading or trailing whitespace
##### Splitting and merging
  - [ ] TC001 - Merge after writing
  - [ ] TC002 - Merge after selection
  - [ ] TC003 - Merge after deleting text
  - [ ] TC004 - Merge after deleting all
  - [ ] TC005 - Merge multiple blocks
  - [ ] TC006 - Splitting/merge list block
#### Undo / Redo - Test Cases
  - [ ] TC001 - Undo/redo block actions
  - [ ] TC002 - Undo/redo text
  - [ ] TC003 - Undo/redo text format

hypest commented:
Note: won't update the branches until we finish with the manual testing, to avoid waiting again for CI everywhere (the gutenberg-mobile side is fully green on CI atm). We can update the branches when we're ready to start merging.

SiobhyB commented:
Noting that I'm seeing strange behaviour when attempting to create more than one new line in the quote block. You'll see from the recording that when I attempt to create a third line, by tapping enter for the second time, the second line is lost and a new block is created:

https://user-images.githubusercontent.com/2998162/110960306-82c79780-8346-11eb-8c67-9cfafcfc710b.mov

There are some existing issues surrounding new lines and the quote block (https://github.com/wordpress-mobile/gutenberg-mobile/issues/2498 and https://github.com/WordPress/gutenberg/issues/27690) but this seems significantly different and isn't something I was able to replicate on my main device (running alpha-280).

I tested this on a Pixel 2 emulator.

SiobhyB commented:
I was able to successfully complete all of the manual tests except for [the multiline components checks](https://github.com/wordpress-mobile/test-cases/blob/trunk/test-cases/gutenberg/writing-flow/multiline-components.md). If I attempt to create more than two new lines with any of the listed blocks, apart from the code block, then text disappears as explained [here](https://github.com/WordPress/gutenberg/pull/29683#issuecomment-797558908) for the quote block. I'm going to do some more testing so that I can verify things work as expected on the main branch for me, just in case the issue is with my local environment, and will update accordingly.

#### General

- [x] TC001 - Paste formatted text copied from website
- [ ] TC001 - Multiline components
    - [ ] Quote block
    - [ ] Verse block
    - [ ] Preformatted block
    - [x] Code block (DEV only)
    - [ ] Pullquote block

#### Rich Text Format

  - [x] TC001 - Bold, Italic, strikethrough buttons
  - [x] TC002 - Alignment buttons
  - [x] TC003 - Alignment Split
  - [x] TC004 - Link button works without selection
  - [x] TC005 - Link button works with a selected word
  - [x] TC006 - Adding a link from a copied URL
  - [x] TC007 - Test format detection under the cursor
  - [x] TC008 - Test formatting doesn't remove leading or trailing whitespace

##### Splitting and merging

  - [x] TC001 - Merge after writing
  - [x] TC002 - Merge after selection
  - [x] TC003 - Merge after deleting text
  - [x] TC004 - Merge after deleting all
  - [x] TC005 - Merge multiple blocks
  - [x] TC006 - Splitting/merge list block
  
#### Undo / Redo - Test Cases

  - [x] TC001 - Undo/redo block actions
  - [x] TC002 - Undo/redo text
  - [x] TC003 - Undo/redo text format

SiobhyB commented:
I confirmed that I'm able to enter multiple lines when using the verse, quote, preformatted, and pullquote blocks on the main branch, as expected. Here's a video of things working as expected with the pullquote block on the main branch:

https://user-images.githubusercontent.com/2998162/110966265-fcfb1a80-834c-11eb-8b75-329d67cfd10c.mov

After I switch to the `rnmobile/try-fix-split-issue-drop-selection-update-event` branch, I have issues when attempting to use more than two lines with all four blocks (verse, quote, preformatted, and pullquote). Here's a recording of what that looks like with the preformatted block:

https://user-images.githubusercontent.com/2998162/110968269-3896e400-834f-11eb-82f5-7ba6aac677ca.mov

In the above example, it took four lines for the block to revert back to the first line (deleting the others that had been created). 

When I initially started testing these, it was happening reliably after two lines had been entered for each block but then began to get a bit harder to replicate, taking a few more lines as the second video shows.

SiobhyB commented:
@jd-alexander pointed me to [the APK here](https://github.com/wordpress-mobile/WordPress-Android/pull/14239#issuecomment-796977633), which I downloaded so that I could go through the tests again on my device. I _wasn't_ able to replicate any of the issues I described, or that aren't already listed as known in the test cases, using it. That's with a Pixel 5 running `version pr-14239-build-99043` and on Android version 11. 

I've only been able to replicate the issues in the Pixel 2 emulator, which is running Android version 10. Interestingly, I'm now no longer unable to reliably replicate the issues there.

I'm not sure if this would be relevant: Stefanos noted that the HTML for the pullquote blocks has a single `<p>` with a `<br>` in the middle, though the testcase says it should be individual `<p>` tags. I've been seeing a mixture of this in my testing, with `<br>` sometimes appearing and sometimes not.

| `<br>` tag  | No `<br>` tag |
| ------------- | ------------- |
| <img src="https://user-images.githubusercontent.com/2998162/110974266-44d26f80-8356-11eb-94f7-5cf5519e19cf.png" width="100%"> | <img src="https://user-images.githubusercontent.com/2998162/110974380-603d7a80-8356-11eb-86bb-558a5f037961.png" width="100%"> |

mkevins commented:
Hi @jd-alexander and @hypest :wave: :smile: 

I recently encountered an issue that I think is related with the split/merge issue (while working on Android's `RichText.onEnter` behavior for the slash-inserter during hackweek). While debugging I noticed a few things that I think are worth mentioning in case it helps for this issue as well, since I have the feeling that many "symptoms" are from the same or similar underlying problem(s).

> I am wondering if this has something to do with how the newline marker is being processed in the `EnterPressedWatcher` I am trying to determine a mechanism for proper diagnosis.

I was testing on a Pixel 3a (physical device) with the Gboard keyboard, and I believe enter key detection is failing in `EnterPressedWatcher`. I believe this is because the Gboard keyboard is actually doing things that we might not normally expect with regard to the `TextWatcher` interface methods.

Here is a table describing some observations that surprised me:

Imagine we have the input with text `Hello worl`, and we then press `d` on the keyboard:

Comparison |`TextWatcher` Interface Method | `text: CharSequence` | `start: Int` | `count: Int` | `after: Int` | `before: Int`
-|-|-|-|-|-|-
Expected | `beforeTextChanged` | "Hello worl" | 10 | 0 | 1 | N/A
Actual | `beforeTextChanged`| "Hello worl" | 6 | 4 | 5 | N/A
Expected | `onTextChanged` | "Hello world" | 10 | 1 | N/A | 0
Actual | `onTextChanged`| "Hello world" | 6 | 5 | N/A | 4

The surprising part was that some characters from the original text were being replaced when I didn't have anything selected (just a normal caret at the end of the input). It seems instead that Gboard is actually replacing the last 4 characters with those same 4 characters _plus_ the typed text. You may even notice a small underline style applied to the characters that will be included in the replacement. Before this, I would have only expected `beforeTextChanged` to be called with a `count` > 0 if there was a selection prior to the new characters being typed (or if characters were being deleted).

Note that the result is the same (and that the sum is still 11 for `start + after` and `start + count` for `beforeTextChanged` and `onTextChanged` respectively).

Interestingly, I tried selecting one of the suggestions offered by Gboard, and the `onEnter` behavior was working normally after that (and the underlined text went away).

I also tried an approach to detect this Gboard behavior and try to deal with that scenario separately. It seems to work from what I tried, so I think it'd be good to get your opinion about the approach used:

<details>
<summary>Gboard EnterPressedWatcher patch</summary>

```diff
diff --git a/packages/react-native-aztec/android/src/main/kotlin/org/wordpress/mobile/ReactNativeAztec/EnterPressedWatcher.kt b/packages/react-native-aztec/android/src/main/kotlin/org/wordpress/mobile/ReactNativeAztec/EnterPressedWatcher.kt
index cc68aac2e1..097ec027fd 100644
--- a/packages/react-native-aztec/android/src/main/kotlin/org/wordpress/mobile/ReactNativeAztec/EnterPressedWatcher.kt
+++ b/packages/react-native-aztec/android/src/main/kotlin/org/wordpress/mobile/ReactNativeAztec/EnterPressedWatcher.kt
@@ -24,6 +24,9 @@ class EnterPressedWatcher(aztecText: AztecText, var enterDeleter: EnterDeleter)
     private var selEnd: Int = 0
     var done = false
 
+    private var gboardReplacement: CharSequence? = null
+
+
     override fun beforeTextChanged(text: CharSequence, start: Int, count: Int, after: Int) {
         val aztecText = aztecTextRef.get()
         if (aztecText?.getAztecKeyListener() != null && !aztecText.isTextChangedListenerDisabled()) {
@@ -32,6 +35,13 @@ class EnterPressedWatcher(aztecText: AztecText, var enterDeleter: EnterDeleter)
             this.start = start
             this.selStart = aztecText.selectionStart
             this.selEnd = aztecText.selectionEnd
+
+            if (selStart == selEnd && 0 < count && count < after) {
+                // possible gboard replacement detected
+                gboardReplacement = text.subSequence(start, start + count)
+            } else {
+                gboardReplacement = null
+            }
         }
     }
 
@@ -40,14 +50,25 @@ class EnterPressedWatcher(aztecText: AztecText, var enterDeleter: EnterDeleter)
         val aztecKeyListener = aztecText?.getAztecKeyListener()
         if (aztecText != null && !aztecText.isTextChangedListenerDisabled() && aztecKeyListener != null) {
             val newTextCopy = SpannableStringBuilder(text)
+
+            var gboardOffset = this.start
+            // If gboard replacement is happening, we offset the start position by the length
+            // of the gboard replacement
+            if (gboardReplacement != null) {
+                val gboardRestored = newTextCopy.subSequence(start, start + before)
+                if (gboardRestored.toString() == gboardReplacement.toString()) {
+                    gboardOffset += gboardRestored.length
+                }
+            }
+
             // if new text length is longer than original text by 1
             if (textBefore?.length == newTextCopy.length - 1) {
                 // now check that the inserted character is actually a NEWLINE
-                if (newTextCopy[this.start] == Constants.NEWLINE) {
+                if (newTextCopy[gboardOffset] == Constants.NEWLINE) {
                     done = false
                     aztecText.editableText.setSpan(EnterPressedUnderway(), 0, 0, Spanned.SPAN_USER)
                     aztecKeyListener.onEnterKey(
-                            newTextCopy.replace(this.start, this.start + 1, ""),
+                            newTextCopy.replace(gboardOffset, gboardOffset + 1, ""),
                             true,
                             selStart,
                             selEnd
@@ -61,9 +82,12 @@ class EnterPressedWatcher(aztecText: AztecText, var enterDeleter: EnterDeleter)
         aztecTextRef.get()?.editableText?.getSpans(0, 0, EnterPressedUnderway::class.java)?.forEach {
             if (!done) {
                 done = true
-                if (enterDeleter.shouldDeleteEnter())
-                    text.replace(start, start + 1, "")
+                if (enterDeleter.shouldDeleteEnter()) {
+                    var gboardOffset = start
+                    gboardReplacement?.let { replacement -> gboardOffset += replacement.length }
+                    text.replace(gboardOffset, gboardOffset + 1, "")
                 }
+            }
             aztecTextRef.get()?.editableText?.removeSpan(it)
         }
     }
```
</details>

Another thing I noticed (though I'm unsure whether it is related to the issues here, or is perhaps expected) is that sometimes a zero-width space was added at the end of the `CharSequence`, but other times it was not present.

**Edit:**

I've opened an issue to track the Gboard behavior issue: https://github.com/wordpress-mobile/gutenberg-mobile/issues/3316

SiobhyB commented:
After some further testing, I found that I was able to replicate the issues I've described on the latest version of the WordPress app, so they're not specific to this branch. The issue with lines sometimes being lost is tricky to reliably replicate, which is what led to the confusion. Sorry for that! I've now created separate issues for these:

* Loss of Previous Lines when Starting Multiple New Lines: https://github.com/WordPress/gutenberg/issues/29861
* Pullquote and Quote Block: Inconsistent Markup for Newlines on Android: https://github.com/WordPress/gutenberg/issues/29862

I've updated [the multiline components test case](https://github.com/wordpress-mobile/test-cases/blob/trunk/test-cases/gutenberg/writing-flow/multiline-components.md) with these two new known issues.

After running through the writing flow test cases again, I confirm that _I wasn't able to find any issues specific to this branch._

hypest commented:
> I recently encountered an issue that I think is related with the split/merge issue

👋 @mkevins, thanks for adding info here! I'm a bit confused though, Are you referring to [the issue (regression actually) at hand](https://github.com/WordPress/gutenberg/issues/29478) for this PR, or some other issue? If other issue, may I suggest moving to a ticket or a different PR and we can discuss there? There are some interesting points you raise in your comment, ultimately related to GBoard suggestion service that has bitten us in the past too 😢 , and probably deserve it's own space.

If your comment is related to this PR, then I'm not sure which split/merge issue you are referring to. Is there some case this fix is missing? Just for context, the underlying issue (previously split text duplicates after merging) is caused by internal data model corruption in the RichText component, itself caused by a selectionChange event that holds obsolete data. I.e. doesn't seem to have something to do with enter-key detection. Let me know if I'm missing something 🤔.

hypest commented:
Thanks for the detailed testing @SiobhyB !

OK, since it looks like there are no specific issues that have been found only occurring in this PR (pending @mkevins' confirmation too) I will go ahead and update the PR and mark it as ready for review.

hypest committed:
Merge branch 'trunk' into rnmobile/try-fix-split-issue-drop-selection-update-event

hypest committed:
Use window.console, to satisfy linter

hypest commented:
👋 @cameronvoell , added you as a reviewer as well as you've worked with Aztec too. @jd-alexander has already given a deep look but it would be nice if you had a look as well to contribute to the code-level confidence of this PR. Let me know if y'all have any thoughts about the review. Thanks!

EDIT: Noticed now that Cameron is away this week so, will remove the review request for the time being.

mkevins commented:
>  @mkevins, thanks for adding info here! I'm a bit confused though, Are you referring to [the issue (regression actually) at hand](https://github.com/WordPress/gutenberg/issues/29478) for this PR, or some other issue? 

Yes, I'm referring to the splitting and merging issue, and that I believe it could be related. Step 4 in the steps to reproduce is:

> Press Enter (Intro) to split the block into two. (if using the simulator, use the device keyboard)

I noticed that @jd-alexander was also observing strange behavior in the underlying `TextWatcher` at `EnterPressedWatcher`, and a closer look at this might reveal a similar or same underlying issue, or at the very least it might help shed light on why some emulators / devices are exhibiting different behavior while testing and debugging these PRs.

> If other issue, may I suggest moving to a ticket or a different PR and we can discuss there? There are some interesting points you raise in your comment, ultimately related to GBoard suggestion service that has bitten us in the past too  , and probably deserve it's own space.

I agree, the gboard suggestion issues probably deserve their own space. I think it may have been useful here as well since these are inherently related due to how we are handling keypresses and text processing in our Aztec integration, and I wanted this discussion to benefit from what I'd recently discovered, in case it helped resolve the differences observed in some testing scenarios.

> If your comment is related to this PR, then I'm not sure which split/merge issue you are referring to. Is there some case this fix is missing? Just for context, the underlying issue (previously split text duplicates after merging) is caused by internal data model corruption in the RichText component, itself caused by a selectionChange event that holds obsolete data. I.e. doesn't seem to have something to do with enter-key detection. Let me know if I'm missing something .

From the backscroll, it seems that for some cases (perhaps due to different emulators or physical devices being used) the issue was not fully resolved (though I'm not sure whether this is still the case :thinking: ). I provided the patch with the intention of demonstrating an approach for the gboard problem, which might address the remaining cases, if there are any.

hypest commented:
Ah, thanks for elaborating @mkevins , I see, much appreciated. My perception at this point is that we can go ahead with the fix unless new issues are found, and the Gboard suggestions proposal can spawn out to a dedicated PR. 👍

jd-alexander commented on packages/react-native-aztec/android/src/main/java/org/wordpress/mobile/ReactNativeAztec/ReactAztecText.java:
```diff
@@ -377,7 +378,22 @@ private void setIntrinsicContentSize() {
+        if (mNativeEventCount < mEventCountSyncFromJS) {
+            // need to sync up to the counter the JS side is expecting. Avoiding setting
+            //   mNativeEventCount directly from the JS side to avoid race conditions, and instead
+            //   syncing just-in-time when we need the new incrementm
```
Minor typo
```suggestion
            //   syncing just-in-time when we need the new increment
```

jd-alexander reviewed (approved):
@hypest I tested this PR again by splitting/merging content in the Paragraph and Heading block and it behaved as expected. I also noticed the `WARN` and `LOG` outputs from the event counter logic in the console when the `onSelectionChange` event is dropped and when `lastEventCount` bumps doesn't occur. 

LGTM 🚢 

- We can keep an eye out for anything suspicious during our next release testing session where we will be publishing the [keep block id on split](https://github.com/WordPress/gutenberg/pull/28505) revert we have been doing. 

- Also, could you please add a description of the changes to the PR since this is the solution we are going with. Thanks much :) 

- Add the respective release notes to this PR and the `gb-mobile` PR as well.

hypest committed:
Fix typo in comments

Co-authored-by: Joel Dean <jdeanjj1000@gmail.com>

hypest committed:
Merge branch 'trunk' into rnmobile/try-fix-split-issue-drop-selection-update-event

hypest committed:
Sync changelog instructions for native mobile

hypest committed:
Changelog entry for the split/merge regression fix

hypest commented:
Thanks for the extra review pass @jd-alexander !

> Also, could you please add a description of the changes to the PR since this is the solution we are going with. Thanks much :)

Sure thing and good reminder! Updated the PR description with a detailed list of the changes and context about them.

> Add the respective release notes to this PR and the gb-mobile PR as well.

Added the lines! 👍

Will merge all PRs when CI gets green across the board!

hypest commented:
All CI is green, including the extended UI tests in the [gutenberg-mobile PR](https://github.com/wordpress-mobile/gutenberg-mobile/pull/3239) and [the WPAndroid PR](https://github.com/wordpress-mobile/WordPress-Android/pull/14239). Will do the merge domino now.
